content/en/blog/2022-08-05-effective-programmers-read-code.md

```diff
@@ -0,0 +1,59 @@
+---
+title: "Effective Programmers Read Code"
+slug: "effective-programmers-read-code"
+date: "2022-08-05T08:10:14-04:00"
+draft: true
+image_webp: images/blog/reading.webp
+image: images/blog/reading.jpg
+author: Benjamin Bengfort
+description: "While we often think ourselves as authors of code we probably spend more time reading code than writing it. Changing our mindset to think of ourselves as avid code readers will make us more effective programmers."
+---
+
+Programmers think of themselves as authors of code but our daily engineering practice probably requires us to spend more time reading code than writing it. Slightly changing our mindset to think of ourselves as avid code readers will make us more effective programmers and ensure that we can quickly create mental maps of our code, a technique that I refer to as the "glass castle".
+
+<!--more-->
+
+When we write source code, we write for two audiences: the execution runtime that the code operates in and other programmers that need to use or maintain the code including future you. Most discussions about how to write good code focus on one or the other of these audiences, however there is much more middle ground between these audiences than those discussions suggest. Both the runtime and the reader have to step through code line by line, interpreting it and possibly jumping to other locations in code to understand the next step.
+
+The operations and jumps described by source code create complex interactions that define the dynamics of our software systems. The only difference between the runtime and the reader is that the runtime handles a long sequence of instructions and the reader builds a mental abstraction of the interactions. As we program, we build this abstraction in our heads and we must mentally hold it in its entirety while we program. Because of the fragility of the model, (even a minor distraction can bring it crashing down) I refer to it as the "glass castle".[^1]
+
+For me, the glass castle is both a structure and a map.[^2] I hold individual named components in mental space such that related components are nearer to each other. The components are linked with mental threads based on their interactions: e.g. calling a method or a function of another component or instantiating and referencing an object. When I mentally zoom into a component I see that it is itself made up of components with interactions to each other. As I code, I mentally zoom in and out of the model getting more detail when I need it and encapsulating and isolating detail when I don't.
```

Comment:
Suggestions: "...in mental space in a way that places related components near one another."
"As I code, I mentally zoom in and out of the model, getting more detail when..."